### DISCUSSION TITLE
[Xenial/Kinetic]  __init__() got an unexpected keyword argument 'strict'

### DISCUSSION THREAD

**Discussion by mintar** [fkie/catkin_lint#93] [General]
### Problem

When **python-future** is installed, catkin_lint crashes with the error above. This is a bug in python-future 0.15 that was [fixed in 0.16](http://python-future.org/whatsnew.html#what-s-new-in-version-0-16-0-2016-10-27), but Xenial still has python-future 0.15.

This is **not a bug in catkin_lint**, but I thought I'd still mention it here so that others with the same problem can find the solution.

### Steps to reproduce

```
$ docker run --rm -it ros:kinetic-ros-core /bin/bash -c 'apt update && apt install -y python-catkin-lint python-future && catkin_create_pkg foo && catkin_lint foo'
# [...]
catkin_lint: internal error: __init__() got an unexpected keyword argument 'strict'
```

### Solution

```bash
sudo apt remove python-future
```

Alternatively, if you really need it, use `pip` to install python-future >= 0.16.

**Comment by mintar** [comment]
I'm closing this immediately because it's a third-party bug, as I said above.

**Reply by roehling** [reply]
I converted this issue into a discussion, which is probably easier to find for people running into this problem.

**Reply by mintar** [reply]
Oh, I didn't even know that GitHub had added a "Discussions" feature. Interesting.

**Reply by roehling** [reply]
It's relatively new and still in beta, so it has to be enabled on a per-repository basis.

**Comment by roehling** [comment]
As this issue seems to be coming up again, I wanted to discuss options here. I am willing to accommodate Xenial with a workaround if it is not too intrusive. Maybe it's sufficient to remove the `strict=True` argument from the constructor, as it is the default setting anyway.

**Reply by jarvisschultz** [reply]
I was unable to remove `python-future`, but I can confirm installing an updated version of `configparser` via `pip` did work around this issue.

@roehling after posting #94 yesterday I did spend a few minutes digging into what it would take to make things work with the old Python 2.7 version of `ConfigParser`, and unfortunately, it's not quite as simple as removing `strict=True` (which is what I was hoping). I believe the usage of `configparser` could be rewritten in a way that is compatible with all versions of `configparser`, but it isn't trivial. I didn't spend too much time on it, but here are a few things I noticed right off the bat:
 - The current code has a lot of snippets like this `config["catkin_lint"]["rosdistro"] = args.rosdistro` that take advantage of a `__setitem__` method newer versions provide. This would need to be replaced with something like `config.set("catkin_lint", "rosdistro", args.rosdistro)` or a `__setitem__` method would need to be provided.
 - The usage of `config.read_file` is not supported with the old version, but I believe `config.readfp` is equivalent
 - This snippet `config[section].update(severity_overrides)` would need to be rewritten to avoid the usage of `__setitem__` or as listed in the first bullet, a `__setitem__` would need to be provided.

**Reply by roehling** [reply]
This pretty much amounts to reverting to the legacy API, which I would really like to avoid. However, the issue has been fixed in Ubuntu 18.04, which still uses the exact same upstream version of python-future (0.15.2), so I backported the fix to Ubuntu 16.04 and published it in my [Ubuntu ROS PPA](https://launchpad.net/~roehling/+archive/ubuntu/ros). Maybe that helps.

**Reply by roehling** [reply]
Here is a direct download link for [python-future_0.15.2-4ubuntu2~16.04_all.deb](https://launchpad.net/~roehling/+archive/ubuntu/ros/+files/python-future_0.15.2-4ubuntu2~16.04_all.deb) if you want to avoid adding the whole PPA to your system.

**Reply by jarvisschultz** [reply]
I agree the effort of reverting to the legacy API isn't worth it, especially since this is a pretty rare circumstance.

Thanks for the deb/PPA definitely could be helpful!
